### ISSUE TITLE
Encode behaviour for constraint violations at a higher level

### ISSUE THREAD

**Issue by samm82** [JacquesCarette/Drasil#3523]
_Note: this is mainly a place for me to collect my thoughts to see if this is something worth pursuing while facilitating conversation about it._

# Context

Currently, information about what to do when a constraint is violated is built by hand in a requirement:

https://github.com/JacquesCarette/Drasil/blob/afb6fb752b8364d2807ced7fc0c1dd6c6aba52b2/code/drasil-example/projectile/lib/Drasil/Projectile/Requirements.hs#L31-L34

This requirement is currently not actually met by the generated program; calculations continue even with constraint violations. (I noticed this while making [this manual test function for invalid inputs](https://github.com/samm82/Drasil/blob/c53404bfe8d2f44ac6d929bbe2aae031d8905c74/code/stable/projectile/projectile_c_p_nol_b_u_v_d/src/python/test/Control_test.py#L37-L53) for my project; note the "xfail" [or "expected failure"].)

I think this choice for the desired behaviour should be captured at a higher level (maybe as a `Choice`?) similarly to #3259 and the related work on generating input FRs (I can't find an issue number); this would allow us to generate the requirement text, the code, and (eventually!) the test cases consistently.

# (**VERY** Preliminary) Proposed Design

I'm thinking the options we could have might be something like:

```haskell
data ConsViolBehav = Error | Warn | Continue
```
where `Error` means "stop the program with an error to the user" (what Projectile _should_ do), `Warn` means "continue the program but warn the user" (what Projectile currently does), and `Continue` means "continue the program as normal". Note that the `Continue` option is essentially the same as not checking the input constraints at all; if we choose to implement this option, I think we can just not generate the requirement at all (and maybe not even display the constraints?) I also think the `Error` option should verify _all_ inputs, detect _all_ errors and output them to the user, _then_ stop execution so the user has the most information (avoiding a back-and-forth process of only correcting one thing at a time).

# Questions

1. Currently, this is what this requirement looks like for Projectile: "Verify-Input-Values: Check the entered input values to ensure that they do not exceed the [data constraints](https://jacquescarette.github.io/Drasil/examples/projectile/SRS/srs/Projectile_SRS.html#Sec:DataConstraints). If any of the input values are out of bounds, an error message is displayed and the calculations stop." The second sentence is really what we're looking to generate here; should this be a separate requirement? If not, should we generate the first sentence as well (since it looks fairly stable) or should we allow the user some customization here?
2.  The current requirement only checks for input constraint violations; should we also be checking for output constraint violations? (This would likely be a separate issue.)

**Comment by samm82** [comment]
I couldn't sleep so figured I'd write my thoughts down 😅 - @balacij This is what I was talking to you about if you're interesting in seeing it "formalized".

**Comment by smiths** [comment]
Sorry you couldn't sleep @samm82.  At least you took advantage of the time! :smile:

This is already a good result for you Masters @samm82.  I like that you found an error in our generator, in the sense that it immediately provides motivation for your work.  (If you haven't already, I strongly suggest that you start a document and add notes to remind you of things you find during your research.  It can be hard to remember in the end.  You might just want to build the infrastructure for your thesis and put bullet points in the sections as things come up.  This particular discussion would go in the motivation section.)

Part of the problem here is that we treat our requirements as display entities only.  They aren't connected to what we do.  My guess is that we could delete the "requirements" from any of our examples and the code would still be generated the same.  I might be wrong about that, but I think we do our generation from the instance models, but it is really the requirements that tell us what instance models are relevant.  The mapping between requirements and instance models is so trivial that we've skipped the connection.  As a consequence, we don't really do anything with requirements that stray from the instance models.  I wonder if there are other examples of this.

For your questions I don't really understand the first one.  The first and second sentence can't really be separated.  "Check the entered input values to ensure that they do not exceed the [data constraints](https://jacquescarette.github.io/Drasil/examples/projectile/SRS/srs/Projectile_SRS.html#Sec:DataConstraints). If any of the input values are out of bounds, an error message is displayed and the calculations stop."  If we just say Check the values, then you haven't said what to do if there is a violation.  If we just say "if any of the input values are out of bounds" we don't have the pointer to the data constraints from the first sentence.

When you are talking about generating the requirement, do you mean generating the text of the requirement, or generating the code that satisfies the requirement?  From the question, I think you mean the first case.  I'm fairly sure that all of our examples have something like the first sentence.  If you did a survey of the examples and found some that didn't match, we could very likely change them to the exact same wording without losing anything.  In that case, I think if the author of the requirements should only have to select that they want a "Verify Inputs" requirement, and the first sentence is automatically given.  The variation would be in the second sentence if we wanted to distinguish between errors (stop the calculations) and warnings.  We could have a parameter that let the author decide whether verify inputs results in warnings or errors.  We should allow for the possibility of two requirements: one for warnings and one for errors.  Each requirement would have to point to a different table of constraints (one for warnings and one for constraints).

@samm82, did I interpret your question 1 correctly?

For your second question, we should also be checking for output constraints.  These can be particularly useful for metamorphic testing where we make sure that an entire suite of tests doesn't violate the constraints.

**Comment by samm82** [comment]
Yes, you interpreted my first question correctly @smiths - I realize now that separating the two sentences doesn't really make sense.

Some other thoughts I had based on our current constraint violation behaviour: 

1. We allow for a different behaviour for physical vs. software constraints, which would likely add some complexity to this requirement (e.g., "Check the entered input values to ensure that they do not exceed the [data constraints](https://jacquescarette.github.io/Drasil/examples/projectile/SRS/srs/Projectile_SRS.html#Sec:DataConstraints). If any of the input values are out of the physical bounds, an error message is displayed and the calculations stop. If any of the input values are out of the software bounds, a warning message is displayed and the calculations continue.")
2. When a single violation is found, an exception is thrown immediately, even if there are other violations. I propose we output all error message _before_ raising an exception so that the user is notified of _all_ violations so they can fix them all at once.

**Comment by JacquesCarette** [comment]
To me, this is a perfect opportunity to dig and generalize. The questions that come to mind:
- what are the different kinds of constraints that can come up?
- where do they come from?
- what are they attached to?
- should they all be checked? [Some invariants would cost as much to check as running the program itself... it would be neat to find some that cost even more]
- what should the process for checking them be? are there choices in this process?
- when the check fails, what to do? What granularity of decision makes sense?

There's probably more, but it will be easier to find them once a stab's been give at the ones that I already list.
